lab_2_tokenize_by_bpe/main.py

```diff
@@ -58,6 +134,38 @@ def train(
     :param num_merges: required number of new tokens
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not (
+        isinstance(word_frequencies, dict)
+        and isinstance(num_merges, int)
+    ):
+        return None
+
+    token_pairs_dict = count_tokens_pairs(word_frequencies)
+    if not token_pairs_dict:
+        return None
+
+    if num_merges > len(token_pairs_dict):
+        num_merges = len(token_pairs_dict)
+
+    max_freq = max(token_pairs_dict.values())
+    max_freq_tokens = []
+
+    for pair in token_pairs_dict:
+        if token_pairs_dict[pair] == max_freq:
+            max_freq_tokens.append(pair)
+
+    max_freq_tokens = sorted(max_freq_tokens, key=lambda x: (-len(''.join(x)), x))
+    token_pair = max_freq_tokens[0]
+
+    word_frequencies = merge_tokens(word_frequencies, token_pair)
+
+    if not word_frequencies:
+        return None
+
+    if num_merges == 1:
+        return word_frequencies
+
+    return train(word_frequencies, num_merges - 1)
```

Comment:
recursion is too much - try to use just `for`

Reply:
done